packages/graphql-server/src/makeMergedSchema/makeMergedSchema.ts

```diff
@@ -70,22 +70,44 @@ const mapFieldsToService = ({
  * If null or invalid value is returned, will trigger a GQL error
  */
 const resolveUnionType = (types: readonly GraphQLObjectType[]) => ({
-  __resolveType(obj: GraphQLObjectType) {
-    // resolves type of object by looking for the largest intersection of common fields
+  __resolveType(obj: Record<string, unknown>) {
+    // if obj has __typname, check that first to resolve type, otherwise, look for largest intersection
+    if (Object.prototype.hasOwnProperty.call(obj, '__typename')) {
+      for (const type of types) {
+        if (type.name === obj['__typename']) {
+          return type.name
+        }
+      }
+    }
     let maxIntersectionType
     let maxIntersectionFields = 0
+    let maxIntersectionIdx = 0
+    const fieldIntersections = new Array(types.length).fill(0)
 
-    for (const type of types) {
+    for (let i = 0; i < types.length; i++) {
+      const type = types[i]
       const fieldIntersection = Object.keys(type.getFields()).filter(
         (field) => field in obj
```

Comment:
In a similar vein to my previous comment, now that I'm thinking about it, should we use `Object.hasOwn` here too or does it not matter?